pkg/ioutils: Make subsequent Close calls a no-op

vvoland opened the pull request:
- related to: https://github.com/moby/moby/pull/46418

Turn subsequent `Close` calls into a no-op and produce a warning with an optional stack trace (if debug mode is enabled).

**- What I did**
Turned subsequent `Close` calls on on (Read/Write/cancel)CloserWrappers into a no-op.

**- How I did it**

**- How to verify it**

**- Description for the changelog**


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented on pkg/pools/pools.go:
```diff
@@ -126,12 +135,23 @@ func (bufPool *BufioWriterPool) Put(b *bufio.Writer) {
+
+func subsequentCloseWarn(name string) {
+	log.G(context.TODO()).Warn("subsequent attempt to close " + name)
+	log.G(context.TODO()).Debugf("stack trace: %s", string(debug.Stack()))
```
Is the `string(debug.Stack()` always evaluated here? (regardless if "debug" logging is enabled)?

If so, we should probably gate this line based on the current debug-level (or would including it in `WithFields()` do the trick?)

vvoland replied:
Unfortunately yes, `WithFields` won't solve this as we need to explicitly convert the `[]byte` to a string, otherwise it will be logged in a binary form (array of bytes like `[255, 0, 1, 5, ...]`).

I can do explicit level check.

thaJeztah replied:
Yeah, maybe we need that; I must say that I didn't look deep into how "costly" this part is, but  I _think_ this code could be in a "hot" path, so it may be relevant if it is.

thaJeztah commented on pkg/pools/pools.go:
```diff
@@ -86,11 +89,17 @@ func Copy(dst io.Writer, src io.Reader) (written int64, err error) {
+			subsequentCloseWarn("ReadCloserWrapper")
+			return nil
+		}
 		if readCloser, ok := r.(io.ReadCloser); ok {
 			readCloser.Close()
 		}
 		bufPool.Put(buf)
+		closed = true
```
Trying to get my head around if this; shouldn't the `closed` var be somehow associated with the reader (`r`)? (as every call to this function could technically pass a completely unrelated `io.Reader`?

vvoland replied:
Each call will have its own `closed`, but yeah that won't help if you do multiple `NewReadCloserWrapper` with the same reader. 
We could also do the check in the `ReadCloserWrapper` directly, but we'd need to make sure if every usage of `NewReadCloserWrapper` would be fine with that.

thaJeztah replied:
we'd need a `NewReadCloserWrapperWrapper` 😂 

I initially was thinking if we need a `sync.Once` for closing (but of course then we'd miss the logs, assuming you wanted those at least for some time to find "bad" uses of closing).

thaJeztah commented on pkg/pools/pools.go:
```diff
@@ -86,7 +90,12 @@ func Copy(dst io.Writer, src io.Reader) (written int64, err error) {
 func (bufPool *BufioReaderPool) NewReadCloserWrapper(buf *bufio.Reader, r io.Reader) io.ReadCloser {
+	var closed int32 = 0
 	return ioutils.NewReadCloserWrapper(r, func() error {
+		if !atomic.CompareAndSwapInt32(&closed, 0, 1) {
```
ha! I was thinking about atomic bool when I started looking (but that led me down the other question I had) 😂

vvoland committed:
pkg/ioutils: Make subsequent Close attempts noop

Turn subsequent `Close` calls into a no-op and produce a warning with an
optional stack trace (if debug mode is enabled).

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

corhere commented on pkg/pools/pools.go:
```diff
@@ -86,7 +90,12 @@ func Copy(dst io.Writer, src io.Reader) (written int64, err error) {
 // NewReadCloserWrapper returns a wrapper which puts the bufio.Reader back
 // into the pool and closes the reader if it's an io.ReadCloser.
 func (bufPool *BufioReaderPool) NewReadCloserWrapper(buf *bufio.Reader, r io.Reader) io.ReadCloser {
+	var closed int32 = 0
 	return ioutils.NewReadCloserWrapper(r, func() error {
+		if !atomic.CompareAndSwapInt32(&closed, 0, 1) {
```
```suggestion
	var closed atomic.Bool
	return ioutils.NewReadCloserWrapper(r, func() error {
		if !closed.CompareAndSwap(false, true) {
```

vvoland replied:
Thanks!

corhere commented on pkg/pools/pools.go:
```diff
@@ -86,7 +90,12 @@ func Copy(dst io.Writer, src io.Reader) (written int64, err error) {
 	return ioutils.NewReadCloserWrapper(r, func() error {
+		if !atomic.CompareAndSwapInt32(&closed, 0, 1) {
+			subsequentCloseWarn("ReadCloserWrapper")
+			return nil
```
[The behavior of Close after the first call is undefined.](https://pkg.go.dev/io#Closer)

Code which calls Close more than once on an arbitrary io.Closer is buggy by definition. Logging a warning is a great first step, especially for code which may discard the error return, but it really should also return an error to maximize fail-loudly-ness without actually panicking.

cpuguy83 replied:
In this case, I would tend to agree, at least with not logging, nobody will ever see this log message message because there generally wouldn't be a reason to correlate that with anything.
If someone does see it it will have very little meaning to them.

If it is an error maybe we'll see it show up in CI?
In most cases if there is a bug I would imagine it would be something like passing a closer to another function and both the outer and the inner have close in a defer.
Not technically correct since the outer function should be delegating the close to the inner function.
That can also get hairy, what if the inner function returns an error? It would be difficult to tell if the closer was closed or not.

Anyway, I see your point that this would be a bug that we should catch, but also possibly it would just as easily be a non-issue.
Not to mention in most places we don't check error on close.

corhere commented on pkg/pools/pools.go:
```diff
@@ -135,3 +149,10 @@ func (bufPool *BufioWriterPool) NewWriteCloserWrapper(buf *bufio.Writer, w io.Wr
 }
+
+func subsequentCloseWarn(name string) {
+	log.G(context.TODO()).Warn("subsequent attempt to close " + name)
```
```suggestion
	log.G(context.TODO()).Error("BUG: subsequent attempt to close " + name)
```
Warnings are for users, IMO. Buggy program behaviour is an error.

cpuguy83 commented on pkg/pools/pools.go:
```diff
@@ -11,9 +11,13 @@ package pools // import "github.com/docker/docker/pkg/pools"
 	"sync"
+	"sync/atomic"
 
+	"github.com/containerd/containerd/log"
```
```suggestion
	"github.com/containerd/log"
```

thaJeztah commented:
@vvoland ptal

thaJeztah reviewed (approved):
LGTM
